p2p/src/sync/peer.rs

```diff
@@ -256,24 +262,39 @@ where
 
         // Check that all the blocks are known and haven't been already requested.
         let ids = block_ids.clone();
-        let best_known_block = self.best_known_block.unwrap_or(0.into());
+        let best_known_block = self.best_known_block.clone();
         self.chainstate_handle
             .call(move |c| {
+                // Check that all blocks are known. Skip the first block as it has already checked.
                 for id in ids {
                     let index = c.get_block_index(&id)?.ok_or(P2pError::ProtocolError(
                         ProtocolError::UnknownBlockRequested(id),
                     ))?;
 
-                    if index.block_height() <= best_known_block {
-                        return Err(P2pError::ProtocolError(
-                            ProtocolError::DuplicatedBlockRequest(id),
-                        ));
+                    if let Some(ref best_known_block) = best_known_block {
+                        if index.block_height() <= best_known_block.block_height() {
+                            // This can be normal in case of reorg, check if the block id is the same.
+                            let known_block_id = c
+                                .get_block_id_from_height(&best_known_block.block_height())?
+                                // This should never fail because we have a block for this height.
+                                .expect("Unable to get block id from height");
+                            if &known_block_id == best_known_block.block_id() {
+                                return Err(P2pError::ProtocolError(
+                                    ProtocolError::DuplicatedBlockRequest(id),
+                                ));
+                            }
+                        }
                     }
                 }
+
                 Result::<_>::Ok(())
             })
             .await??;
 
+        // A peer can ignore the headers request if it is in the initial block download state.
+        // Assume this is the case if it asks us for blocks.
+        self.last_activity = None;
```

Comment:
In the definition of `last_activity` in the struct, please write why it can be None. Without context it's really hard to guess what's going on. I'd even prefer an enum to handle all that...